### ISSUE TITLE
WebAppReporter Out of Memory Error

### ISSUE THREAD

**Issue by CleanHit** [oss-review-toolkit/ort#2288]
Running this command

```bash
JAVA_OPTS="-Xms1024M -Xmx6g -XX:MaxMetaspaceSize=2g -Dkotlin.daemon.jvm.options='-Xmx2g'" ort --info report -f NoticeByPackage,StaticHtml,WebApp -i ./here-oss/evaluation/v1/evaluation-result.yml -o ./here-oss/report/v1
``` 
gives me:
```bash
Created 'NoticeByPackage' report:
        /here-oss/report/v1/NOTICE_BY_PACKAGE
Created 'StaticHtml' report:
        /here-oss/report/v1/scan-report.html
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.Arrays.copyOf(Arrays.java:3746)
        at java.base/java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:172)
        at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:538)
        at java.base/java.lang.StringBuilder.append(StringBuilder.java:174)
        at java.base/java.lang.StringBuilder.append(StringBuilder.java:85)
        at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:573)
        at java.base/java.lang.StringBuilder.append(StringBuilder.java:204)
        at java.base/java.lang.StringBuilder.append(StringBuilder.java:85)
        at kotlin.text.StringsKt__StringBuilderKt.appendElement(StringBuilder.kt:59)
        at kotlin.sequences.SequencesKt___SequencesKt.joinTo(_Sequences.kt:1724)
        at kotlin.sequences.SequencesKt___SequencesKt.joinToString(_Sequences.kt:1743)
        at kotlin.sequences.SequencesKt___SequencesKt.joinToString$default(_Sequences.kt:1742)
        at kotlin.text.StringsKt__StringsJVMKt.replace(StringsJVM.kt:76)
        at kotlin.text.StringsKt__StringsJVMKt.replace$default(StringsJVM.kt:75)
        at com.here.ort.reporter.reporters.WebAppReporter.generateReport(WebAppReporter.kt:43)
        at com.here.ort.commands.ReporterCommand.run(ReporterCommand.kt:152)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:139)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:14)
        at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:221)
        at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:218)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:236)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:259)
        at com.here.ort.OrtMainKt.main(OrtMain.kt:205)
```
I have 32GB of RAM and I can see that the parameters are being passed to the java execution. I had similar issues with `analyzer` and `scanner` and I've solved it by adding
```gradle
org.gradle.jvmargs=-Xms1024M -Xmx6g -XX:MaxMetaspaceSize=2g -Dkotlin.daemon.jvm.options="-Xmx2g"
```
To the `gradle.properties` file of the scanned project. Since `ort report` doesn't use the setting in this file I had to pass the variables as command line parameters.

**Comment by sschuberth** [comment]
Please just disable the Web App reporter for now. It's known to be feature-incomplete (e.g. excludes are not properly displayed) and to be slow for big projects.

**Comment by tsteenbe** [comment]
@CleanHit  Can you please try to recreate this issue with the latest SHA1 of ORT. We merged today a [major write of the WebApp reporter](https://github.com/heremaps/oss-review-toolkit/commit/4c6836e5ff0cb3dfeba992a099391dc33fbad720)

**Comment by CleanHit** [comment]
@tsteenbe I can try it only next week.

**Comment by CleanHit** [comment]
@tsteenbe I've tyied it again with commit `ad4f85` and now I'm getting a new Java related error: 
```bash
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.LinkedHashMap.newNode(LinkedHashMap.java:256)
        at java.base/java.util.HashMap.putVal(HashMap.java:626)
        at java.base/java.util.HashMap.put(HashMap.java:607)
        at com.fasterxml.jackson.databind.node.ObjectNode.replace(ObjectNode.java:455)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:284)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeArray(JsonNodeDeserializer.java:437)
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:261)
        at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:68)
        at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:15)
        at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4173)
        at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:2577)
        at com.fasterxml.jackson.databind.ObjectMapper.valueToTree(ObjectMapper.java:2966)
        at com.here.ort.reporter.model.EvaluatedModel.toSortedTree(EvaluatedModel.kt:162)
        at com.here.ort.reporter.model.EvaluatedModel.toJson(EvaluatedModel.kt:153)
        at com.here.ort.reporter.reporters.WebAppReporter.generateReport(WebAppReporter.kt:45)
        at com.here.ort.commands.ReporterCommand.run(ReporterCommand.kt:163)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:139)
        at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:14)
        at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:221)
        at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:218)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:236)
        at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:259)
```
I've also noticed that my Java process related to the execution don't die. I have 45 open Process, each process uses 4.9% MEM and 0% CPU.

My Set-up is:
```
OS: Fedora 3
JAVA_HOME = <path_to_openJDK>/jdk-12-0-2
ANDROID_HOME = <path_to_SDK>/AndroidSDK
```

**Comment by sschuberth** [comment]
That's probably something for @mnonnenmacher to look at. Maybe we can write out the evaluated model in chunks?

**Comment by mnonnenmacher** [comment]
@sschuberth This happens while sorting the evaluated model (`com.here.ort.reporter.model.EvaluatedModel.toSortedTree(EvaluatedModel.kt:162)`), at this point the model needs to be in memory. Writing is already done using streaming.

@CleanHit As your machine has 32GB RAM, please try increasing the value for `-Xmx`.

**Comment by CleanHit** [comment]
@mnonnenmacher Thanks! Unfortunately, I've put _ORT_ on hold after #2314 passes the internal review and gets implemented, since it's crucial for my work. When that happens I'll be able to check your suggestion in this issue.

**Comment by CleanHit** [comment]
> @CleanHit As your machine has 32GB RAM, please try increasing the value for `-Xmx`.

@mnonnenmacher do yo mean the _Java JVM_ -Xmx or the `-Dkotlin.daemon.jvm.options="-Xmx"`? I set both.

**Comment by mnonnenmacher** [comment]
@CleanHit The `-Xmx` for the JVM.

**Comment by CleanHit** [comment]
@mnonnenmacher Setting `-Xmx=8g` generated the WebAppReporter.
